File: Python/substrings-that-begin-and-end-with-the-same-letter.py
# Time:  O(n)
# Space: O(1)

import collections


class Solution(object):
    def numberOfSubstrings(self, s):
        """
        :type s: str
        :rtype: int
        """
        result = 0
        cnt = collections.Counter()
        for c in s:
            cnt[c] += 1
            result += cnt[c]
        return result


# Time:  O(n)
# Space: O(1)
import collections


class Solution(object):
    def numberOfSubstrings(self, s):
        """
        :type s: str
        :rtype: int
        """
        return sum(v*(v+1)//2 for v in collections.Counter(s).itervalues())

File: C++/substrings-that-begin-and-end-with-the-same-letter.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    long long numberOfSubstrings(string s) {
        unordered_map<char, int> cnt;
        int64_t result = 0;
        for (const auto& c : s) {
            result += ++cnt[c];
        }
        return result;
    }
};

// Time:  O(n)
// Space: O(1)
class Solution2 {
public:
    long long numberOfSubstrings(string s) {
        unordered_map<char, int64_t> cnt;
        
        for (const auto& c : s) {
            ++cnt[c];
        }
        int64_t result = 0;
        for (const auto& [_, v] : cnt) {
            result += v * (v + 1) / 2;
        }
        return result;
    }
};
